File: packages/adapter-opencode/src/opencode-adapter.ts
/**
 * OpenCode SDK Adapter
 *
 * 透過 @opencode-ai/sdk 呼叫 OpenCode 執行任務。
 * OpenCode 使用 Client/Server 架構，需先啟動 server 再透過 HTTP API 操作。
 * 支援 session create、prompt、fork、headless mode。
 */

import { createOpencode } from "@opencode-ai/sdk";
import type { OpencodeClient } from "@opencode-ai/sdk";
import type {
  AgentAdapter,
  AgentType,
  ExecuteOptions,
  Task,
  TaskResult,
} from "@dev-autopilot/core";

/**
 * OpenCode SDK Adapter
 *
 * 將 dev-autopilot 的 Task 轉換為 OpenCode SDK 的 session.prompt 呼叫。
 */
export class OpenCodeAdapter implements AgentAdapter {
  readonly name: AgentType = "opencode";

  private client: OpencodeClient | null = null;
  private serverClose: (() => void) | null = null;

  /**
   * 初始化 OpenCode SDK（啟動 server + 建立 client）
   */
  private async ensureClient(): Promise<OpencodeClient> {
    if (this.client) return this.client;

    const { client, server } = await createOpencode();
    this.client = client;
    this.serverClose = server.close;
    return client;
  }

  /**
   * 執行單一任務
   *
   * 1. 建立或接續 session
   * 2. 發送 prompt（task spec）
   * 3. 等待完成，提取結果
   *
   * @param task - 要執行的任務
   * @param options - 執行選項
   * @returns 任務執行結果
   */
  async executeTask(task: Task, options: ExecuteOptions): Promise<TaskResult> {
    const startTime = Date.now();

    try {
      const client = await this.ensureClient();

      // 建立或 fork session
      let sessionId: string;
      if (options.sessionId && options.forkSession) {
        const forkResult = await client.session.fork({
          path: { id: options.sessionId },
        });
        const forkData = forkResult.data as { id: string } | undefined;
        sessionId = forkData?.id ?? options.sessionId;
      } else if (options.sessionId) {
        sessionId = options.sessionId;
      } else {
        const createResult = await client.session.create({
          query: { directory: options.cwd },
        });
        const createData = createResult.data as { id: string } | undefined;
        sessionId = createData?.id ?? "";
      }

      options.onProgress?.(`[${task.id}] OpenCode session: ${sessionId}`);

      // 構建 prompt
      const prompt = this.buildPrompt(task);

      // 發送 prompt
      const promptResult = await client.session.prompt({
        path: { id: sessionId },
        body: {
          parts: [{ type: "text", text: prompt }],
        },
      });

      const promptData = promptResult.data as {
        info?: { cost?: number; error?: unknown };
      } | undefined;

      const cost = promptData?.info?.cost ?? 0;
      const hasError = !!promptData?.info?.error;

      return {
        task_id: task.id,
        session_id: sessionId,
        status: hasError ? "failed" : "success",
        cost_usd: cost,
        duration_ms: Date.now() - startTime,
        verification_passed: !hasError,
        error: hasError ? String(promptData?.info?.error) : undefined,
      };
    } catch (error) {
      return {
        task_id: task.id,
        status: "failed",
        duration_ms: Date.now() - startTime,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  /**
   * 檢查 OpenCode CLI 是否可用
   *
   * 嘗試執行 `opencode --version` 確認 CLI 已安裝。
   */
  async isAvailable(): Promise<boolean> {
    try {
      const { execFile } = await import("node:child_process");
      const { promisify } = await import("node:util");
      const execFileAsync = promisify(execFile);
      await execFileAsync("opencode", ["--version"]);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * 恢復指定 session
   *
   * @param sessionId - 要恢復的 session ID
   */
  async resumeSession(sessionId: string): Promise<void> {
    const client = await this.ensureClient();
    await client.session.get({
      path: { id: sessionId },
    });
  }

  /**
   * 關閉 OpenCode server
   */
  async dispose(): Promise<void> {
    this.serverClose?.();
    this.client = null;
    this.serverClose = null;
  }

  /**
   * 構建送給 OpenCode 的 prompt
   */
  private buildPrompt(task: Task): string {
    let prompt = `請執行以下任務：\n\n## ${task.title}\n\n${task.spec}`;

    if (task.verify_command) {
      prompt += `\n\n## 驗收條件\n執行完成後請用以下指令驗證：\n\`\`\`bash\n${task.verify_command}\n\`\`\``;
    }

    return prompt;
  }
}

File: python/devap/adapters/opencode_adapter.py
"""
OpenCode SDK Adapter

透過 HTTP API 呼叫 OpenCode 執行任務。
OpenCode 使用 Client/Server 架構，需先啟動 server 再透過 HTTP API 操作。
"""

from __future__ import annotations

import asyncio
import json
import time
from typing import Any, Optional

from devap.models.types import (
    AgentAdapter,
    ExecuteOptions,
    Task,
    TaskResult,
    TaskStatus,
)


class OpenCodeAdapter(AgentAdapter):
    """
    OpenCode SDK Adapter

    透過 OpenCode CLI 的 HTTP API 執行任務。
    """

    def __init__(self) -> None:
        self._server_url: Optional[str] = None

    @property
    def name(self) -> str:
        """agent 類型名稱"""
        return "opencode"

    async def execute_task(
        self, task: Task, options: ExecuteOptions
    ) -> TaskResult:
        """
        執行單一任務

        透過 opencode CLI 子進程執行，解析 JSON 輸出。

        Args:
            task: 要執行的任務
            options: 執行選項

        Returns:
            任務執行結果
        """
        start_time = time.monotonic()

        try:
            prompt = self._build_prompt(task)
            args = self._build_args(task, options)

            proc = await asyncio.create_subprocess_exec(
                "opencode", *args,
                cwd=options.cwd,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            stdout, stderr = await proc.communicate(input=prompt.encode())
            stdout_str = stdout.decode()

            if proc.returncode != 0 and not stdout_str.strip():
                return TaskResult(
                    task_id=task.id,
                    status="failed",
                    duration_ms=(time.monotonic() - start_time) * 1000,
                    error=f"opencode 退出碼 {proc.returncode}: {stderr.decode()}",
                )

            return self._parse_result(task.id, stdout_str, start_time)

        except Exception as e:
            return TaskResult(
                task_id=task.id,
                status="failed",
                duration_ms=(time.monotonic() - start_time) * 1000,
                error=str(e),
            )

    async def is_available(self) -> bool:
        """
        檢查 OpenCode CLI 是否可用

        嘗試執行 `opencode --version` 確認已安裝。
        """
        try:
            proc = await asyncio.create_subprocess_exec(
                "opencode", "--version",
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            await proc.communicate()
            return proc.returncode == 0
        except (FileNotFoundError, OSError):
            return False

    async def resume_session(self, session_id: str) -> None:
        """恢復指定 session"""
        pass

    def _build_prompt(self, task: Task) -> str:
        """構建送給 OpenCode 的 prompt"""
        prompt = f"請執行以下任務：\n\n## {task.title}\n\n{task.spec}"

        if task.verify_command:
            prompt += (
                f"\n\n## 驗收條件\n"
                f"執行完成後請用以下指令驗證：\n"
                f"```bash\n{task.verify_command}\n```"
            )

        return prompt

    def _build_args(self, task: Task, options: ExecuteOptions) -> list[str]:
        """構建 CLI 啟動參數"""
        args = ["-p", "--output-format", "json"]

        if options.session_id:
            args.extend(["--resume", options.session_id])

        if task.max_turns:
            args.extend(["--max-turns", str(task.max_turns)])

        return args

    def _parse_result(
        self,
        task_id: str,
        stdout: str,
        start_time: float,
    ) -> TaskResult:
        """解析 CLI 輸出為 TaskResult"""
        try:
            parsed: dict[str, Any] = json.loads(stdout.strip())
        except (json.JSONDecodeError, ValueError):
            return TaskResult(
                task_id=task_id,
                status="failed",
                duration_ms=(time.monotonic() - start_time) * 1000,
                error=f"OpenCode 輸出解析失敗：{stdout[:200]}",
            )

        session_id = parsed.get("session_id")
        cost_usd = parsed.get("cost_usd", 0.0)
        duration_ms = parsed.get("duration_ms")
        is_error = parsed.get("is_error", False)
        subtype = parsed.get("subtype", "")

        if not is_error and subtype == "success":
            return TaskResult(
                task_id=task_id,
                session_id=str(session_id) if session_id else None,
                status="success",
                cost_usd=float(cost_usd) if cost_usd is not None else None,
                duration_ms=float(duration_ms) if duration_ms is not None else (
                    (time.monotonic() - start_time) * 1000
                ),
                verification_passed=True,
            )

        status: TaskStatus = (
            "timeout"
            if subtype in ("error_max_turns", "error_max_budget_usd")
            else "failed"
        )

        return TaskResult(
            task_id=task_id,
            session_id=str(session_id) if session_id else None,
            status=status,
            cost_usd=float(cost_usd) if cost_usd is not None else None,
            duration_ms=float(duration_ms) if duration_ms is not None else (
                (time.monotonic() - start_time) * 1000
            ),
            error=subtype if status != "success" else None,
        )
